lib/ProductOpener/Paths.pm

```diff
@@ -0,0 +1,310 @@
+# This file is part of Product Opener.
+#
+# Product Opener
+# Copyright (C) 2011-2023 Association Open Food Facts
+# Contact: contact@openfoodfacts.org
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::Paths - functions to get, check and initialize important directories
+
+=head1 SYNOPSIS
+
+…
+
+=cut
+
+package ProductOpener::Paths;
+
+use ProductOpener::PerlStandards;
+use Exporter qw< import >;
+
+use Log::Any qw($log);
+
+BEGIN {
+	use vars qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+		%BASE_DIRS
+
+		&base_paths
+		&check_missing_dirs
+		&ensure_dir_created
+		&ensure_dir_created_or_die
+	);    # symbols to export on request
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+}
+
+use vars @EXPORT_OK;
+
+use ProductOpener::Config qw/:all/;
+
+=head1 VARIABLES
+
+=cut
+
+=head2 %BASE_DIRS
+A hashmap containing references to base directories
+=cut
+
+%BASE_DIRS = ();
```

Comment:
I like the idea of having extra documentation about what each directory is, but on the other hand, it's less obvious when reading the code that a directory is for data (input or output), or that it will be publicly accessible on the web. It was more clear to me with $data_root and $www_root. The extra indirection kind of obfuscate things, especially as the mapping from DIR_NAME to the actual name of the directory on the disk varies.

Reply:
Maybe the extra documentation could be put in README.md files instead.
And we could better arrange what we have under $data_root, to clearly separate input and output data.

Reply:
The main idea was not that but to forbid the dynamic creation of the base folders (creating folders inside base folders is ok, but not creating those base folders). It's to avoid creating data at the wrong place (eg. on the VM disk instead of a ZFS dataset).

You can see that apache start up script verifies all these folders exists or refuse to start (with an explicit message).

I am also boldly in favor of a $data_root and $cache_root, but I didn't want to change everything right now.

Maybe we could prefix folders name, or have more explicit name (we don't use them so many times in the code, so longer names are affordable). At least, add PUBLIC to every public folders.